src/parse.rs

```diff
@@ -692,15 +744,134 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn try_consume_named_capture_group_name(&mut self) -> Option<String> {
+        let mut cursor = self.input.clone();
+
+        match cursor.next() {
+            Some('<') => {}
+            _ => return None,
+        }
+
+        let mut group_name = String::new();
+
+        if let Some(c) = cursor.next() {
+            if c.is_id_start() || c == '$' || c == '_' {
+                group_name.push(c);
+            } else {
+                return None;
+            }
+        } else {
+            return None;
+        }
+
+        loop {
+            match cursor.next() {
+                Some('>') => break,
+                Some(c) => {
+                    if c.is_id_continue() || c == '$' || c == '_' || c == '\u{200C}' /* <ZWNJ> */ || c == '\u{200D}'
+                    /* <ZWJ> */
+                    {
+                        group_name.push(c);
+                    } else {
+                        return None;
+                    }
+                }
+                None => return None,
+            }
+        }
+
+        self.input = cursor;
+
+        Some(group_name)
+    }
+
+    fn try_consume_named_capture_group_backreference(&mut self) -> Result<Option<u32>, Error> {
+        let mut cursor = self.input.clone();
+
+        match cursor.next() {
+            Some('<') => {}
+            _ => return Ok(None),
+        }
+
+        let mut group_name = String::new();
+
+        if let Some(c) = cursor.next() {
+            if c.is_id_start() || c == '$' || c == '_' {
+                group_name.push(c);
+            } else {
+                return error(format!("Invalid character in capture group name: {}", c));
+            }
+        } else {
+            return Ok(None);
+        }
+
+        loop {
+            match cursor.next() {
+                Some('>') => break,
+                Some(c) => {
+                    if c.is_id_continue() || c == '$' || c == '_' || c == '\u{200C}' /* <ZWNJ> */ || c == '\u{200D}'
+                    /* <ZWJ> */
+                    {
+                        group_name.push(c);
+                    } else {
+                        return error(format!("Invalid character in capture group name: {}", c));
+                    }
+                }
+                None => return Ok(None),
+            }
+        }
+
+        if let Some(index) = self.named_group_indicies.get(&group_name) {
+            self.input = cursor;
+            Ok(Some(*index))
+        } else if self.flags.unicode {
+            error(format!(
+                "Backreference to invalid named capture group: {}",
+                &group_name
+            ))
+        } else {
+            Ok(None)
+        }
+    }
+
+    // Quickly parse all capture groups.
+    fn parse_capture_groups(&mut self) {
```

Comment:
> Is the idea pre-populate `named_group_indices` for backreferences?

Yes that is the idea. The challange with doing a pass on the IR is, that without the `unicode` flag, a `\k` can be parsed as the literal `k` character if there are no named capture groups in the entire regex. I actually did not implement the unicode flag handling correctly before. I just pushed aa3d2ac and fixed that error while adding the AnnexB tests for the named capture groups. If you take a look at the new `\k` handling in `consume_atom_escape` ([line 746+](https://github.com/ridiculousfish/regress/commit/aa3d2ac9681ffdd17c1a1d2ceb3becfc7d5010a8#diff-5583e398a8deec11b274d9965eff8b5ade5226c7a020ca214a27d7d07dcc8a29R746)) you may see why I went with this approach.

Concerning the correctness and the approach in general; while implementing I took a look at the v8 implementation. They choose this approach too. While I'm not fluent in c++, I think my parsing function should be identical to the v8 one. See the v8 version here: https://github.com/v8/v8/blob/master/src/regexp/regexp-parser.cc#L728

Because of the wired AnnexB edge cases, I'm not 100% sure if the IR pass approach would be that easy to implement. But I'm also only getting stared on this codebase, so I would like to defer to your judgement on this ;)

Reply:
Nicely found. If v8 implements it this way it's probably fine. I really do think there's a bug with \ handling in character classes: [this specific line](https://github.com/v8/v8/blob/4e1a6c418ad333ad8d2ddccad443bd86ce932c0f/src/regexp/regexp-parser.cc#L751) is the logic that's missing in this diff and I think `[\](.)]` should trigger it. Once that's fixed it should be OK.

Reply:
I will do some testing with that example and fix it.

With b444806 I removed the `boa_unicode` dependency and refactored the `gen-folds` crate into a more general `gen-unicode` crate. You might want to take a look at that commit specifically to see the difference.

I'm also working on implementing the unicode escape parsing more closely to the spec. I will ping you once all that work is done.

Reply:
Indeed you where right, I did miss the escape handling. I fixed it and wrote a test.

Unicode escape parsing is now done in the main parsing. I think that should conclude all the issues you found.